### PR TITLE
Add loss and OOO impairment

### PR DESCRIPTION
**Pull request by elv-serban**
Bonus:
  - cfg structure to store configuration (not fully populated)
  - ctx structure to store running data

Example:

./udpreplay -m "5000:5:8000:8" -h 127.0.0.1 -p 12000  sample.pcap

### CODE REVIEW

File: src/udpreplay.cpp [hunk 1]
```diff
@@ -12,10 +12,167 @@
...
+ * Not thread safe (uses globals)
+ */
+static int pkt_queue(ctx_t ctx, const u_char *d, ssize_t len, sockaddr_in addr) {
+
+    cfg_t *cfg = ctx.cfg;
+
+    int rc;
+    struct pkt_t pkt = {d, len, addr};
+
+    static int64_t lossint_count = cfg->loss_int_sec; // packets until next loss
```
Comment (JanFalkin) [inline src/udpreplay.cpp]: Just wondering why we are declaring the locals as static?  We want them not on the stack?

File: src/udpreplay.cpp [hunk 2]
```diff
@@ -12,10 +12,167 @@
...
+    pcfg->loss_pkts = atoi(ptr);
+    ptr = strtok(NULL, ":");
+    if (ptr == NULL)
+        return -1;
+    pcfg->ooo_int_sec = atoi(ptr);
+    ptr = strtok(NULL, ":");
+    if (ptr == NULL)
+        return -1;
+    pcfg->ooo_pkts = atoi(ptr);
+
```
Comment (JanFalkin) [inline src/udpreplay.cpp:168]: if atoi fails you will get 0 which will appear to be a valid ts. Perhaps a warn on 0 may be appropriate?

File: src/udpreplay.cpp [hunk 3]
```diff
@@ -100,10 +268,18 @@ int main(int argc, char *argv[]) {
...
            "  -p          destination port (optional)\n"
+           "  -m          impairment (format: LOSS_INT_SEC:N_LOSS:OOO_INT_SEC:N_OOO)\n"
+           "\n"
+           "Example: ./udpreplay -m 5000:5:8000:8 -h 127.0.0.1 -p 12000 sample.pcap\n"
         << std::endl;
     return 1;
   }
 
+  if (parse_impairment(&cfg, impairment) < 0) {
+      std::cerr << "bad impairment options: " << impairment << std::endl;
```
Comment (JanFalkin) [inline src/udpreplay.cpp:287]: change of convention using std and operator << instead of printf?

**Review by JanFalkin — APPROVED** [review]
The only true substantive change would be a warning for a failing atoi

File: src/udpreplay.cpp (hunk 1 above)
Comment (elv-serban) [reply src/udpreplay.cpp]: These are globals.  I need to move them in the context `ctx_t`

File: src/udpreplay.cpp (hunk 2 above)
Comment (elv-serban) [reply src/udpreplay.cpp:168]: I added a check - if both loss and OOO intervals are 0 I return an error (bad impairment).

File: src/udpreplay.cpp (hunk 3 above)
Comment (elv-serban) [reply src/udpreplay.cpp:287]: The original upstream code uses cout/cerr.  I'll have to change the printfs to that but will do that in a separate commit - too much busywork.

File: src/udpreplay.cpp (hunk 1 above)
Comment (elv-serban) [reply src/udpreplay.cpp]: Ok I moved them to ctx_t

**Pull request merged**